src/test/java/org/sonarsource/scanner/jenkins/pipeline/WaitForQualityGateStepTest.java

```diff
@@ -0,0 +1,257 @@
+      public void evaluate() throws Throwable {
+        handler.status = "SUCCESS";
+        handler.analysisId = FAKE_ANALYSIS_ID;
+        handler.qgStatus = "OK";
+        WorkflowRun b = submitPipeline();
+
+        story.j.assertBuildStatusSuccess(story.j.waitForCompletion(b));
+      }
+    });
+  }
+
+  @Test
+  public void finishEarlyKo() {
+    story.addStep(new Statement() {
+      @Override
+      public void evaluate() throws Throwable {
+        handler.status = "SUCCESS";
+        handler.analysisId = FAKE_ANALYSIS_ID;
+        handler.qgStatus = "KO";
+        WorkflowRun b = submitPipeline();
+
+        story.j.assertBuildStatus(Result.FAILURE, story.j.waitForCompletion(b));
+      }
+    });
+  }
+
+  @Test
+  public void waitForQualityGateOk_survive_restart() {
+    story.addStep(new Statement() {
+      @Override
+      public void evaluate() throws Throwable {
+        handler.status = "PENDING";
+        submitPipeline();
+        waitForStepToWait();
+      }
+    });
+    story.addStep(new Statement() {
+      @Override
+      public void evaluate() throws Throwable {
+        WorkflowJob p = story.j.jenkins.getItemByFullName(JOB_NAME, WorkflowJob.class);
+        WorkflowRun b = p.getLastBuild();
+        submitWebHook("SUCCESS", "OK", b);
+        story.j.assertBuildStatusSuccess(story.j.waitForCompletion(b));
+      }
+    });
+  }
+
+  @Test
+  public void waitForQualityGateOk_recheck_ws_on_restart() {
+    story.addStep(new Statement() {
+      @Override
+      public void evaluate() throws Throwable {
+        handler.status = "PENDING";
+        submitPipeline();
+        waitForStepToWait();
+      }
+    });
+    story.addStep(new Statement() {
+      @Override
+      public void evaluate() throws Throwable {
+        handler.status = "SUCCESS";
+        handler.analysisId = FAKE_ANALYSIS_ID;
+        handler.qgStatus = "OK";
+      }
+    });
+    story.addStep(new Statement() {
+      @Override
+      public void evaluate() throws Throwable {
+        WorkflowJob p = story.j.jenkins.getItemByFullName(JOB_NAME, WorkflowJob.class);
+        WorkflowRun b = p.getLastBuild();
+        story.j.assertBuildStatusSuccess(story.j.waitForCompletion(b));
+      }
+    });
+  }
+
+  private void submitWebHook(String endTaskStatus, String qgStatus, WorkflowRun b) throws InterruptedException, IOException, SAXException {
+    waitForStepToWait();
+
+    SonarAnalysisAction action = b.getAction(SonarAnalysisAction.class);
+    assertThat(action.getCeTaskId()).isEqualTo(FAKE_TASK_ID);
+    story.j.postJSON("sonarqube-webhook/", "{\n" +
+      "\"taskId\":\"" + FAKE_TASK_ID + "\",\n" +
+      "\"status\":\"" + endTaskStatus + "\",\n" +
+      "\"qualityGate\":{\"status\":\"" + qgStatus + "\"}\n" +
+      "}");
+  }
+
+  private void waitForStepToWait() throws InterruptedException {
+    // Wait for the step to register to the webhook listener
+    while (SonarQubeWebHook.get().listeners.isEmpty()) {
+      Thread.sleep(500);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Remove this use of "Thread.sleep()". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2925)